src/hyperlight_host/src/hypervisor/hyperlight_vm.rs

```diff
@@ -0,0 +1,768 @@
+
+        fn read_addrs(
+            &mut self,
+            mut gva: u64,
+            mut data: &mut [u8],
+            dbg_mem_access_fn: Arc<Mutex<dyn DbgMemAccessHandlerCaller>>,
+        ) -> crate::Result<()> {
+            let data_len = data.len();
+            log::debug!("Read addr: {:X} len: {:X}", gva, data_len);
+
+            while !data.is_empty() {
+                let gpa = self.vm.translate_gva(gva)?;
+
+                let read_len = std::cmp::min(
+                    data.len(),
+                    (PAGE_SIZE - (gpa & (PAGE_SIZE - 1))).try_into().unwrap(),
+                );
+                let offset = (gpa as usize)
+                    .checked_sub(SandboxMemoryLayout::BASE_ADDRESS)
+                    .ok_or_else(|| {
+                        log::warn!(
+                            "gva=0x{:#X} causes subtract with underflow: \"gpa - BASE_ADDRESS={:#X}-{:#X}\"",
+                            gva, gpa, SandboxMemoryLayout::BASE_ADDRESS);
+                        HyperlightError::TranslateGuestAddress(gva)
+                    })?;
+
+                dbg_mem_access_fn
+                    .try_lock()
+                    .map_err(|e| new_error!("Error locking at {}:{}: {}", file!(), line!(), e))?
+                    .read(offset, &mut data[..read_len])?;
+
+                data = &mut data[read_len..];
+                gva += read_len as u64;
+            }
+
+            Ok(())
+        }
+
+        /// Copies the data from the provided slice to the guest memory address
+        /// The address is checked to be a valid guest address
+        fn write_addrs(
+            &mut self,
+            mut gva: u64,
+            mut data: &[u8],
+            dbg_mem_access_fn: Arc<Mutex<dyn DbgMemAccessHandlerCaller>>,
+        ) -> crate::Result<()> {
+            let data_len = data.len();
+            log::debug!("Write addr: {:X} len: {:X}", gva, data_len);
+
+            while !data.is_empty() {
+                let gpa = self.vm.translate_gva(gva)?;
+
+                let write_len = std::cmp::min(
+                    data.len(),
+                    (PAGE_SIZE - (gpa & (PAGE_SIZE - 1))).try_into().unwrap(),
+                );
+                let offset = (gpa as usize)
+                .checked_sub(SandboxMemoryLayout::BASE_ADDRESS)
+                .ok_or_else(|| {
+                    log::warn!(
+                        "gva=0x{:#X} causes subtract with underflow: \"gpa - BASE_ADDRESS={:#X}-{:#X}\"",
+                        gva, gpa, SandboxMemoryLayout::BASE_ADDRESS);
+                    HyperlightError::TranslateGuestAddress(gva)
+                })?;
+
+                dbg_mem_access_fn
+                    .try_lock()
+                    .map_err(|e| new_error!("Error locking at {}:{}: {}", file!(), line!(), e))?
+                    .write(offset, data)?;
+
+                data = &data[write_len..];
+                gva += write_len as u64;
+            }
+
+            Ok(())
+        }
+    }
+}
```

Comment:
Maybe @dblnz wants to have a look at this?

Reply:
This is copied from previous implementation. Git doesn't indicate it's the same file because more than 50% changed

Reply:
Given the scope of the changes I think we should have @dblnz  review this